### PR TITLE
Add different error measures

### PR DESCRIPTION
**Pull request by szymon159**

### CODE REVIEW

File: error_functions.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    def cross_entropy(calculated: float, expected: float, derivative: bool = False) -> float:
+        if derivative:
+            return expected / calculated - (1 - expected) / (1 - calculated)
+        else:
+            return -expected * log(calculated) - (1 - expected) * log(1 - calculated)
+
+    @staticmethod
+    def mean_squared(calculated: float, expected: float, derivative: bool = False) -> float:
+        if derivative:
+            return expected - calculated
```
Comment (Michalski-Artur) [inline error_functions.py]: Are you sure about derivative here? I would expect it should rather be something similar to described here: https://towardsdatascience.com/gradient-descent-from-scratch-e8b75fa986cc

File: neuron.py [hunk 2]
```diff
@@ -1,41 +1,45 @@
...
     # Based on: https://machinelearningmastery.com/implement-backpropagation-algorithm-scratch-python/
     def calculate_error(self, expected_output: list[float], next_layer) -> float:
         error = 0
         if next_layer is None:
-            error = self.__output - expected_output[self.neuron_index_in_layer]
+            error = -self.__error_function(self.__output, expected_output[self.neuron_index_in_layer], True)
         else:
             for neuron in next_layer.neurons:
-                error += neuron.weights[self.neuron_index_in_layer] * neuron.delta
+                error = neuron.weights[self.neuron_index_in_layer] * neuron.delta
```
Comment (Michalski-Artur) [inline neuron.py]: I think I don't get it. Why error should depend only on error propagated throught last neuron in the next layer?

File: neuron.py [hunk 3]
```diff
@@ -1,41 +1,45 @@
...
-        self.__output = self.__activation_function(dot(self.weights, input_value))
+        self.__output: float = self.__activation_function(dot(self.weights, input_value))
         return self.__output
 
     # Based on: https://machinelearningmastery.com/implement-backpropagation-algorithm-scratch-python/
     def calculate_error(self, expected_output: list[float], next_layer) -> float:
         error = 0
         if next_layer is None:
-            error = self.__output - expected_output[self.neuron_index_in_layer]
+            error = -self.__error_function(self.__output, expected_output[self.neuron_index_in_layer], True)
```
Comment (Michalski-Artur) [inline neuron.py]: Where does this minus before calculating error function comes from? This value is later also multiplied by `-1` when updating weights. If this is because results get better then I have a sense maybe there is an issue with error function derivative, but it is just a guess

**Review by Michalski-Artur — CHANGES_REQUESTED** [review]

File: neuron.py (hunk 2 above)
Comment (szymon159) [reply neuron.py]: Typo

File: neuron.py (hunk 3 above)
Comment (szymon159) [reply neuron.py]: Minus is necessary but I've moved it to functions definitions

File: error_functions.py (hunk 1 above)
Comment (szymon159) [reply error_functions.py]: I think so: if out function is `f(x,y) = 1/2 * (x-y)^2 = 1/2 (x^2 - 2xy + y^2)`, then derivative `(df(x,y)/x) = 1/2 * (2x - 2y) = x - y` and the same for `(df(x,y)/dy)`

File: network.py [hunk 4]
```diff
@@ -33,38 +34,38 @@ def add_network_layer(self, layer: Layer) -> None:
 
     def train(self, training_set: pd.DataFrame, visualize: bool = False) -> None:
         current_iteration = 0
-        while not self.__stop_condition_met(current_iteration):
+        mean_error = 1.0
```
Comment (Michalski-Artur) [inline network.py]: I think it would be better to initalize mean error with `self.__error_threshold`. It would also prevent from any single epoch not running when error threshold is set to more than 1

**Review by Michalski-Artur — APPROVED** [review]

File: error_functions.py (hunk 1 above)
Comment (Michalski-Artur) [reply error_functions.py]: Yep, you are right :)

File: network.py (hunk 4 above)
Comment (szymon159) [reply network.py]: Good idea, done

File: network.py [hunk 5]
```diff
@@ -28,37 +34,38 @@ def add_network_layer(self, layer: Layer) -> None:
 
     def train(self, training_set: pd.DataFrame, visualize: bool = False) -> None:
         current_iteration = 0
-        while not self.__stop_condition_met(current_iteration):
+        mean_error = self.__error_threshold
+        while not self.__stop_condition_met(current_iteration, mean_error):
             current_iteration += 1
-            sum_error = 0
+            if current_iteration == self.__epoch_no:
```
Comment (Michalski-Artur) [inline network.py:40]: Why we need this? Isn't this part of stop_condition_met method?

File: network.py [hunk 6]
```diff
@@ -28,37 +34,38 @@ def add_network_layer(self, layer: Layer) -> None:
...
             input_value: np.ndarray
             expected_result: float
+            sample_len = len(x_train)
             for (input_value, expected_result) in zip(x_train, y_train):
                 outputs = self.__forward_pass(input_value)
-                expected = [0 for i in range(len(outputs))]
+                expected = [0 for _ in range(len(outputs))]
                 expected[int(expected_result) - 1] = 1
-                sum_error += sum([(expected[i] - outputs[i]) ** 2 for i in range(len(expected))]) # TODO: This error probably doesn't correspond with calculated one?
+                mean_error += self.__error_function(outputs[int(expected_result) - 1], 1) / sample_len
```
Comment (Michalski-Artur) [inline network.py:52]: Maybe divide by `sample_len` after the loop? Or it leads to overflow or other issues?

File: neuron.py [hunk 7]
```diff
@@ -41,4 +45,4 @@ def calculate_error(self, expected_output: list[float], next_layer) -> float:
 
     def update_weights(self, learning_rate: float) -> None:
         delta_weights = np.multiply(self.__input, self.delta * learning_rate)
-        self.weights = np.add(self.weights, -delta_weights)
+        self.weights = np.clip(np.add(self.weights, -delta_weights), -10, 10)
```
Comment (Michalski-Artur) [inline neuron.py:48]: As we discussed clipping is problematic especially for the regression, but probably also for classification. Maybe we can at least clip depending on problem type until we find better solution?

**Review by Michalski-Artur — APPROVED** [review]

File: network.py (hunk 5 above)
Comment (szymon159) [reply network.py:40]: That was for debug (to put breakpoint in last epoch), removed in some other branch

File: neuron.py (hunk 7 above)
Comment (szymon159) [reply neuron.py:48]: Will figure out later :)

File: network.py (hunk 6 above)
Comment (szymon159) [reply network.py:52]: Will do but in other branch to avoid conflicts
